packages/signature-controller/src/SignatureController.ts

```diff
@@ -413,19 +417,26 @@ export class SignatureController extends BaseControllerV2<
       ...(version && { version }),
     };
 
-    const signaturePromise = messageManager.waitForFinishStatus(
-      messageParamsWithId,
-      messageName,
-    );
-
+    let resultCallbacks: AcceptResultCallbacks | undefined;
     try {
-      await this.#requestApproval(messageParamsWithId, approvalType);
+      const acceptResult = await this.#requestApproval(
+        messageParamsWithId,
+        approvalType,
+      );
+      resultCallbacks = acceptResult.resultCallbacks;
     } catch (error) {
       this.#cancelAbstractMessage(messageManager, messageId);
       throw ethErrors.provider.userRejectedRequest(
         'User rejected the request.',
       );
     }
+
+    const signaturePromise = messageManager.waitForFinishStatus(
```

Comment:
If we're now doing this after we've already handled the rejection, won't this promise never resolve in the case of rejection as the event will have already fired before creating the listener?

Perhaps we could leave the `waitForFinishStatus` where it was, but instead await the `signaturePromise` using a try catch where we can intercept errors and call `resultCallbacks.error`?

Reply:
> If we're now doing this after we've already handled the rejection, won't this promise never resolve in the case of rejection as the event will have already fired before creating the listener?

Promise will never be created by `waitForFinishStatus` if rejection happens by `#requestApproval` because we throw.

Also if we await `signaturePromise` before `#requestApproval` as we do it in the past code, it will never get resolved at that point because signing operation will never be finished then it will stuck. 

Besides, I forgot to call `resultCallbacks.error` in the `#requestApproval` rejection, just saw that now.

Reply:
Sure, that makes sense in order to decouple that result callbacks, since approval created by `SignatureController`

Reply:
Done